### ISSUE TITLE
Part of the Firefox44 or later of compatibility issues, problems with the variable scope (e.g. userContext is not defined).

### ISSUE THREAD

**Issue by mnphnic** [vimperator/vimperator-labs#422]
## TL;DR

We should not use the `userContext`?

<hr/>

Sorry for my super-poor English.
But, this problem is serious.
Until now `userContext` had pointing to the global-scope (not chrome-window-object).
However, from the Firefox44 `userContext` it is no longer referenced.
The reason is obvious.
It is because tailored to the ES6 global-lexical-bindings rules.
So, `userContext` that are const declared globally are no longer exported to the global scope.

```javascript
// https://github.com/vimperator/vimperator-labs/blob/master/common/content/liberator.js#L13
const userContext = { __proto__: modules };
```

As a temporary workaround, I'm using the following code (this is user-side script).
```javascript
try {
  liberator.eval('userContext');
} catch (e) {
  var userContext = liberator.eval('this');
}
```
EDIT: more better temporary workaround.
```javascript
if (typeof userContext === 'undefined')
  var userContext = liberator.eval('this');
```

`this` scope of the above is consistent with global scope that are referenced in the command-line (same as the former `userContext`).

In the first place, leakage of `userContext` to the global scope has originally does not assume?

## References
* [Breaking changes in let and const in Firefox Nightly 44 | Mozilla Add-ons Blog](https://blog.mozilla.org/addons/2015/10/14/breaking-changes-let-const-firefox-nightly-44/)

**Comment by mnphnic** [comment]
Oh, I was mistaken.
`userContext` can be replaced with a variable declaration.

old
```javascript
userContext.polyfill = [];
```

new
```
var polyfill = [];
```

Was self resolved...

**Comment by gkatsev** [comment]
Probably what needs to happen is to use the `Object.defineProperty` workaround suggested in the post linked. Would you be up to making a PR for this? Thanks.

**Comment by mnphnic** [comment]
Thanks for replay.
I want to confirm, does there is no problem in using the userContext in user-side script?

**Comment by gkatsev** [comment]
I am not sure whether it's should be accessible to the user but it probably should be accessible within other files in vimperator. Plus, it was previously accessible to the user, so, I don't think it would be bad if it user accessible.

**Comment by mnphnic** [comment]
I think so too.

If I have to PR, it will be the code such as the following.

```javascript
var evalContext = liberator.eval('this');
Object.defineProperty(evalContext, 'userContext', {
  value: evalContext,
  enumerable: true,
  writable: true
});
```

it is right?

**Comment by gkatsev** [comment]
Couldn't you just do:
```js
Object.defineProperty(this, 'userContext', {
  value: { __proto__: modules },
  enumerable: true,
  writeable: true
});
```

**Comment by mnphnic** [comment]
Gotcha!
Global scope of the user-side script is the same as the `liberator.plugins.contexts[PATH]`.
However, it global scope of the Vimperator module-side was the same as the `liberator.eval('this')`.
I see...

**Comment by mnphnic** [comment]
Sorry, my slowest commit.
and sorry, I make dead commit object.

**Comment by mnphnic** [comment]
I've found a more this problems.
```javascript
// from https://github.com/vimperator/vimperator-labs/blob/master/common/content/util.js

const {XHTML, XUL, NS} = (function () {
    function Namespace(prefix, uri) {
        this.prefix = prefix;
        this.uri = uri;
    }
    Namespace.prototype.toString = function toString() { return this.uri; };
    return {
        XHTML: new Namespace("html", "http://www.w3.org/1999/xhtml"),
        XUL  : new Namespace("xul", "http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"),
        NS   : new Namespace("liberator", "http://vimperator.org/namespaces/liberator"),
    };
})();
```
I think many of the user-plugins is using the `NS`.
I rewrote the patch. and more better than previous patch.
